### PR TITLE
Add Service Account support

### PR DESCRIPTION
**Pull request by tim-oster**
This PR adds support for using an 1Password Service Account token as opposed to using a Connect Server setup (#79).

The implementation is best effort but most likely not ideal. It is a wrapper around the `op` CLI to be able to make use of the service account token. As of today, there seems to be no other option to interface with 1password other than the cli or a connect server.

The biggest issue in this implementation is, that updating an item will first delete it and then createa new one. The main reason for this is, that `op` does not support updating items using json. Ideally, we would use `FroceNew` for this in the provider but since it needs to be compatible with the Connect Server mode, this is not possible.

### CODE REVIEW

**Comment by franklouwers** [comment]
Does this mean that when updating an item, it will get a new UUID and new urls?

**Comment by tim-oster** [comment]
@franklouwers yes, unfortunately.

It might be tweakable to use the actual `op item edit` command. The only downside I see is that it might not be fully compatible with everything that the json schema offers and that it would essentially leak all credentials of the command as arguments ([more info](https://developer.1password.com/docs/cli/reference/management-commands/item/#item-edit-flags)).

**Comment by gabrielrinaldi** [comment]
@tim-oster does that mean the agent running this needs the `op` cli installed? We are using Terraform Cloud, which I think would not support that unfortunately. But excited to see progress on support for SAs.

**Comment by volodymyrZotov** [comment]
@tim-oster Great contribution! Your work will help us a lot to release this feature even faster! 🚀
I'll review your pr shortly.

File: onepassword/cli/op.go
Comment (volodymyrZotov) [inline onepassword/cli/op.go:87]: I'd not go with deleting/creating the item as you'll need to operate with new `uuid` each time you re-created it and keeping in mind Service Accounts [quotas and rate limits](https://developer.1password.com/docs/service-accounts/get-started#rate-limits).
But I see the reason you implemented it in this way.

I'd like to have the ability to simply update the existing one and use a similar implementation as for  `ItemCreate`. I'll raise that within the team to see what we can do with that.

File: onepassword/cli/exec.go
Comment (tim-oster) [reply onepassword/cli/exec.go]: The distinction was made to in attempt to keep the internal workings clutter out of the public interface defition, especially with all the flag wrappers and so on. But I see your point, refactored.

File: onepassword/cli/op.go
Comment (tim-oster) [reply onepassword/cli/op.go:92]: Totally agree, I don't like the way it is implemented either at atm. If we could get a json edit flow, that would be amazing.

**Comment by tim-oster** [comment]
@volodymyrZotov thanks for the review!

@gabrielrinaldi yes, unfortunately using the CLI seems to be the only option at the moment. I am not too familiar with tf cloud, do you have access to a local file system? In that case, it might be worth investigating if we can bundle the cli binary together with the provider binary. Though this would drastically increase the provider's binary size.

**Comment by volodymyrZotov** [comment]
@gabrielrinaldi [Only Install Standalone Binaries](https://developer.hashicorp.com/terraform/cloud-docs/run/install-software#only-install-standalone-binaries) article might be helpful to understand how you can add additional binaries.

And this one [Install 1Password CLI](https://developer.1password.com/docs/cli/get-started/#step-1-install-1password-cli) might help to install. See `Linux -> apt`, or whatever is suitable for you.

Hope those will help you to figure out how to do that  😃 Let me know how it's going for you, please.

**Comment by gabrielrinaldi** [comment]
@volodymyrZotov amazing, thanks for pointing it out! Can't wait for this to be merged.

**Comment by volodymyrZotov** [comment]
Hi folks! And especially @tim-oster 👋 Just want to put an update about the item edit. 

Implementing that with the current cli version brings some difficulties, so **we decided to add a piped input feature to the cli `item edit` command**, which would allow us to pass the `Item` struct to perform the updates. So with the newer cli version you will be able to implement `EditItem` function in the same way as `CreateItem`.

I'm on that feature right now. I'll keep you posted about the progress. The initial aim is to have a new cli release with this feature later next week, but will see how it goes. 🚀

**Comment by tim-oster** [comment]
@volodymyrZotov amazing, thanks for having a look at this so quickly! I'll update the PR as soon the CLI is released. Looking forward to getting this merged 💪

**Comment by volodymyrZotov** [comment]
Hi @tim-oster. Just an update on that. We're still working on the **item edit piped input feature**. I'll let you know as soon as we have a new cli release with it and happy to help bring your PR to the finish line 🙂

**Comment by maxpowis-bp** [comment]
I just started a complex implementation an realize I might just be stuck at this stage due to this bug. Any ETA for the release ?

**Comment by speier** [comment]
In the meantime here is a hack using `op` cli locally..

```hcl
variable "item_ref" {}

locals {
  temp_dir = "${path.root}/op_tmp"
}

resource "local_file" "output" {
  filename = "${local.temp_dir}/${timestamp()}.txt"
  content  = ""
}

resource "null_resource" "op_read" {
  triggers = {
    once = timestamp()
  }
  provisioner "local-exec" {
    command = "op read ${var.item_ref} > ${local_file.output.filename}"
  }
}

data "local_file" "output" {
  depends_on = [null_resource.op_read]
  filename   = local_file.output.filename
}

resource "null_resource" "deleteLocalFile" {
  depends_on = [null_resource.op_read, data.local_file.output]
  triggers = {
    once = timestamp()
  }
  provisioner "local-exec" {
    command = "rm -rf ${local.temp_dir}"
  }
}

output "item_value" {
  value = data.local_file.output.content
}
```

usage example:

```hcl
module "op" {
  source   = "../modules/op"
  item_ref = "op://myvault/fooprovider/api-token"
}

output "result" {
  value = module.op.item_value
}
```

**Comment by maxpowis-bp** [comment]
Thanks for the quick answer, I am using terraform-cloud though; probably won't do the trick :/

**Comment by hollow** [comment]
Having to fall back to calling `op` from Terraform either through `local-exec` or from the provider code sounds like the wrong direction for this feature and will prevent Terraform Cloud users from using it.

**Comment by speier** [comment]
Sure, totally agree guys, `local-exec` is not the way. It's just a temp hack. Ideally a proper provider should work with 1Password API.

**Comment by volodymyrZotov** [comment]
@tim-oster Hi. Some updates about **`op item edit` piped input feature**.
I know that your PR is blocked by that, but want to let you know that we're actively working on that feature on our side. 
I'll drop a message here once it's released. Stay tuned. 😃

**Comment by volodymyrZotov** [comment]
@tim-oster Hi! 👋 We plan to release a new CLI version that includes the feature to edit items using templates later this week. 
Would you be able to continue the work on the PR? 😃

**Comment by dbrennand** [comment]
@tim-oster @volodymyrZotov - Looks like the new CLI version was released 🙂 https://app-updates.agilebits.com/product_history/CLI2#v2230001

**Comment by tim-oster** [comment]
@volodymyrZotov @dbrennand the new CLI update is awesome! Works as expected!

**Comment by dbrennand** [comment]
> @volodymyrZotov @dbrennand the new CLI update is awesome! Works as expected!

Awesome 🙌🏻 - hopefully this gets merged soon! 😃

**Comment by volodymyrZotov** [comment]
@tim-oster Hi. I pushed some commits to slightly change the error messages and address a couple of issues (see comments above).

While testing I found at least one issue that we need to address before moving forward. It can't create passwords when using CLI and throws an error:
```
[ERROR] 2023/11/21 17:32:33 unable to process line 1: Validation: (validateVaultItem failed to Validate), Couldn't validate the item: "[ItemValidator] has found 1 errors, 0 warnings: \nErrors:{1. Password item requires ps value}"
```

I'll continue looking into that tomorrow and continue testing to see if I can find any other issues.

**Review by volodymyrZotov — APPROVED** [review]

**Comment by volodymyrZotov** [comment]
Added myself as a co-author to sign all the commits.

**Pull request merged**
